### ISSUE TITLE
free mem by madvise(MADV_DONTNEED) may not work when enabled linux transparent hugepage

### ISSUE THREAD

**Issue by jianpanlanyue** [gperftools/gperftools#1073]
On linux(before kernel version 4.4.0) with transparent hugepage, tcmalloc interface 'heap stats' show a lot memory has released to OS, but command 'top' and '/proc/smaps' show process still uses it. This often happens, and leading to high mem usage.  On system with no transparent hugepage or disable hugepage, it works well as expectly.

Does tcmalloc has other way to release memory?

**Comment by jianpanlanyue** [comment]
I found, without tcmalloc, only use native mmap() interface,  high memory usage can appear easily:
environment:  
  1.kernel 4.4.0 on x86_64
  2.enable thp:
    echo always > /sys/kernel/mm/transparent_hugepage/enable
    echo always > /sys/kernel/mm/transparent_hugepage/defrag
    echo 2000000 > /sys/kernel/mm/transparent_hugepage/khugepaged/pages_to_scan ( faster defrag pages to reproduce problem)

problem: 
  1. use mmap() to allocate 4096 bytes for 1024*512 times (4096*1024*512=2G).
  2. use madvise(MADV_DONTNEED) to free most of the above pages, but reserve a few pages(by if（i%33==0) continue;), then process's physical memory firstly come down, but after a few seconds, it rise back to 2G again, and can't come down forever.
  3. if i delete this condition(if（i%33==0) continue;) or disable transparent_hugepage by setting 'enable' and 'defrag' to never, all go well and the physical memory can come down expectly.

test code:

#include <stdio.h>
#include <memory.h>
#include <stdlib.h>
#include <sys/mman.h>
#include <errno.h>
#include <assert.h>

#define PAGE_SIZE 4096
#define PAGE_COUNT 1024*512
int main()
{
  void** table = (void**)malloc(sizeof(void*) * PAGE_COUNT);
  printf("begin mmap...\n");

  for (int i=0; i<PAGE_COUNT; i++) {
    table[i] = mmap(NULL, PAGE_SIZE, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1 ,0);
    assert(table[i] != MAP_FAILED);
    memset(table[i], 1, PAGE_SIZE);
  }
  
  printf("mmap ok, press enter to free most of them\n");
  getchar();

  //it behaves not expectly: after most pages freed, thp make it rise to 2G again
  for(int i=0; i<PAGE_COUNT; i++) {
    if (i%33==0) continue;
    if (madvise(table[i], PAGE_SIZE, MADV_DONTNEED) != 0)
      printf("madvise error, errno:%d\n", errno);
  }

  printf("release finish\n");
  free(table);
  getchar();
  getchar();
}

**Comment by alk** [comment]
Yes, and this known issue. This is arguably kernel bug. This is because default value of /sys/kernel/mm/khugepaged/max_ptes_none is set to 511. Setting it to 0 will fix it.

With value set to 511 khugepaged will convert chunks of 2 megs to huge page even it just 1 small page is used there (rest is madvise-dont-need-ed).

Another way to work around is to disable returning memory back to kernel. Set TCMALLOC_RELEASE_RATE to 0. Of course it means that some kinds of memory fragmentation where we previously at least returned fragmented memory back to kernel, will now actually fragment memory badly.

**Comment by victorstewart** [comment]
have you all any experience with how these effects manifest when running many vastly overcommitting processes on the same machine? reduction in TLB misses is great and all in theory but i'm concerned about extreme memory pressure resulting in OOMM process kills.

I wrote a little program to play around with different madvise strategies. (had to rename the extension from .cpp to .txt so github would let me upload it lol). trivially compilable and runnable like so..

```
g++ mem.test.cpp -o mem.test
./mem.test 100000 MADV_COLD
```
can try `MADV_DONTNEED`, `MADV_FREE`, `MADV_COLD`, `MADV_PAGEOUT`

i'm focusing on how long these operations take, and how many resident pages the process still has when it completes (I assumed there would be 0 since I was memset-ing every page to 0... and claimed elsewhere that replaces pages with the zero page.... but clearly not?). 

with /sys/kernel/mm/khugepaged/max_ptes_none set the default 511...

```
root@clr-18a47c16c91c437c8ce854a37e6f5e26~ # ./mem.test 100000 MADV_DONTNEED
Memory Space 1.93TB
Resident 13.07GB
3066 ms
```
then setting /sys/kernel/mm/khugepaged/max_ptes_none to 0...

root@clr-18a47c16c91c437c8ce854a37e6f5e26~ # ./mem.test 100000 MADV_DONTNEED
Memory Space 1.93TB
Resident 10.49GB
2560 ms

but THEN... when i outright disable transparent hugepages....

```
root@clr-18a47c16c91c437c8ce854a37e6f5e26~ # ./mem.test 100000 MADV_DONTNEED
Memory Space 1.93TB
Resident 0.01GB
389 ms
```

[mem.test.txt](https://github.com/gperftools/gperftools/files/6438424/mem.test.txt)

**Comment by alk** [comment]
Sorry, I don't undestand the question(s) in latest comments. Please elaborate. Otherwise assuming this ticket is about max pte none thingy being borked by default.
